fred2/missioncutscenesdlg.cpp

```diff
@@ -0,0 +1,466 @@
+	DDX_Text(pDX, IDC_CUTSCENE_NAME, m_name);
+	DDX_Text(pDX, IDC_CUTSCENE_HELP_BOX, m_desc);
+	//}}AFX_DATA_MAP
+	DDV_MaxChars(pDX, m_name, NAME_LENGTH - 1);
+}
+
+BEGIN_MESSAGE_MAP(CMissionCutscenesDlg, CDialog)
+	//{{AFX_MSG_MAP(CMissionCutscenesDlg)
+	ON_CBN_SELCHANGE(IDC_DISPLAY_CUTSCENE_TYPES_DROP, OnSelchangeDisplayCutsceneTypesDrop)
+	ON_NOTIFY(TVN_SELCHANGED, IDC_CUTSCENES_TREE, OnSelchangedCutscenesTree)
+	ON_NOTIFY(NM_RCLICK, IDC_CUTSCENES_TREE, OnRclickCutscenesTree)
+	ON_NOTIFY(TVN_ENDLABELEDIT, IDC_CUTSCENES_TREE, OnEndlabeleditCutscenesTree)
+	ON_NOTIFY(TVN_BEGINLABELEDIT, IDC_CUTSCENES_TREE, OnBeginlabeleditCutscenesTree)
+	ON_BN_CLICKED(IDC_BUTTON_NEW_CUTSCENE, OnButtonNewCutscene)
+	ON_CBN_SELCHANGE(IDC_CUTSCENE_TYPE_DROP, OnSelchangeCutsceneTypeDrop)
+	ON_EN_CHANGE(IDC_CUTSCENE_NAME, OnChangeCutsceneName)
+	ON_EN_CHANGE(IDC_CUTSCENE_HELP_BOX, OnChangeCutsceneName)
+	ON_BN_CLICKED(ID_OK, OnOk)
+	ON_WM_CLOSE()
+	//}}AFX_MSG_MAP
+END_MESSAGE_MAP()
+
+/////////////////////////////////////////////////////////////////////////////
+// CMissionCutscenesDlg message handlers
+
+// Initialization: sets up internal working copy of mission cutscenes and cutscene trees.
+void CMissionCutscenesDlg::load_tree()
+{
+	m_cutscenes_tree.select_sexp_node = select_sexp_node;
+	select_sexp_node = -1;
+
+	m_cutscenes_tree.clear_tree();
+	m_cutscenes.clear();
+	m_sig.clear();
+	for (int i=0; i<(int)The_mission.cutscenes.size(); i++) {
+		m_cutscenes.push_back(The_mission.cutscenes[i]);
+		m_sig.push_back(i);
+
+		if (m_cutscenes[i].filename[0] == '\0')
+			strcpy_s(m_cutscenes[i].filename, "<Unnamed>");
+
+		m_cutscenes[i].formula = m_cutscenes_tree.load_sub_tree(The_mission.cutscenes[i].formula, true, "true");
+	}
+
+	m_cutscenes_tree.post_load();
+	cur_cutscene = -1;
+	update_cur_cutscene();
+}
+
+// create the CTreeCtrl tree from the cutscene tree, filtering based on m_display_cutscene_types
+void CMissionCutscenesDlg::create_tree()
+{
+	m_desc = _T(cutscene_descriptions[m_display_cutscene_types]);
+
+	m_cutscenes_tree.DeleteAllItems();
+	m_cutscenes_tree.reset_handles();
+	for (int i = 0; i < (int)m_cutscenes.size(); i++) {
+		if (m_cutscenes[i].type != m_display_cutscene_types)
+			continue;
+
+		HTREEITEM h = m_cutscenes_tree.insert(m_cutscenes[i].filename);
+		m_cutscenes_tree.SetItemData(h, m_cutscenes[i].formula);
+		m_cutscenes_tree.add_sub_tree(m_cutscenes[i].formula, h);
+	}
+
+	cur_cutscene = -1;
+	update_cur_cutscene();
+}
+
+// Display cutscene types selection changed, so update the display
+void CMissionCutscenesDlg::OnSelchangeDisplayCutsceneTypesDrop()
+{
+	UpdateData(TRUE);
+	create_tree();
+}
+
+// New tree item selected.  Because cutscene info is displayed for the selected tree item,
+// we need to update the display when this occurs.
+void CMissionCutscenesDlg::OnSelchangedCutscenesTree(NMHDR* pNMHDR, LRESULT* pResult)
+{
+	NM_TREEVIEW* pNMTreeView = (NM_TREEVIEW*)pNMHDR;
```

Comment:
I would try a reinterpret cast here, but I'm not going to require it.  I know this comes with the territory of FRED code.